### ISSUE TITLE
Missing grad_fn information while torch.compile with customized gradient function

### ISSUE THREAD

**Issue by Taot-chen** [pytorch/pytorch#131974]
### 🐛 Describe the bug

I was using torch.compile to optimize a transpose single operator network, and instead of triggering the output_type configuration of alias_of_input, I wanted to use fw_outs directly as the output of the network, so I tried to add a custom gradient function to the network in anticipation of triggering the output_type configuration of custom_function_view. I used the following script as described earlier,
`# *_*coding:utf-8 *_*
import torch
import torch.nn as nn
import torch.nn.functional as F
import json
import numpy as np
import torch._dynamo as dynamo

class CustomFunction(torch.autograd.Function):
    # @staticmethod
    def forward(self, input):
        return torch.transpose(input, 0, 1)

class CustomTranspose_2D(nn.Module):
    def __init__(self):
        super().__init__()
    def forward(self, input):
        return CustomFunction.apply(input)

if __name__ == '__main__':
    device = 'cpu'
    backend = 'cpu'
    shape2D = (64, 128)
    input_data = np.arange(1.0, shape2D[0] * shape2D[1] + 1).astype(np.float32)
    input_data = input_data.reshape(shape2D)
    mul1 = torch.ones(128, 64).to(device)
    input_data = torch.tensor(input_data , requires_grad=True).to(device)
    print("input_data: ", input_data)
    
    model = CustomTranspose_2D().to(device)
    optimized_model = torch.compile(model, backend=backend)
    
    
    output = optimized_model(input_data)
    print("output: ", output)
`
Inside torch._functorch._aot_autograd.collect_metadata_analysis, the value obtained for grad_fn is None, as shown below
![578651757f66e21a4420d355eecfe50b](https://github.com/user-attachments/assets/c0c6bcb2-257d-4cec-8226-2f13d4a5dbcb)
Therefore, the output_type configuration of alias_of_input is still triggered.
Under such conditions, inside torch._functorch._aot_autograd.functional_utils.gen_alias_from_base there will be in-place processing of fw_outs, which is not what I expect.
<img width="551" alt="dcc62f9066c877c667822e754d2d0f28" src="https://github.com/user-attachments/assets/5a882ae2-5d01-4979-aa99-2028c5f0f8d5">

Is there something wrong with my previous processing? What should I do in order for the output_type to be set to **custom_function_view** or **non_alias**?

### Versions

env:
Python  3.11.9
torch     2.3.0+cpu
numpy  1.26.3

**Comment by yf225** [comment]
cc. @bdhirsh

**Comment by masnesral** [comment]
@yf225 / @bdhirsh I'm helping with the issue scrubbing this quarter. I can't tell if we just never really responded here or if there's an actual bug to fix.

**Comment by bdhirsh** [comment]
@Taot-chen can you describe more what your goal is? 

One surprising thing I am seeing from your example is that if you print the dynamo output (running with `TORCH_LOGS="graph_code"`, it looks like dynamo **eliminates** the custom `autograd.Function`, and replaces it with the view:
```
 ===== __compiled_fn_2 =====
 /data/users/hirsheybar/a/pytorch/torch/fx/_lazy_graph_module.py class GraphModule(torch.nn.Module):
    def forward(self, L_input_: "f32[64, 128][128, 1]cpu"):
        l_input_ = L_input_

         # File: /data/users/hirsheybar/a/pytorch/tmp6.py:12 in forward, code: return torch.transpose(input, 0, 1)
        transpose: "f32[128, 64][1, 128]cpu" = torch.transpose(l_input_, 0, 1);  l_input_ = None
        return (transpose,)
```

That would explain why we aren't hitting the custom view logic in AOTAutograd. @zou3519 do you know why dynamo wouldn't be preserving the custom autograd.Function here?

To be fair, though, it looks like it is semantics preserving in this case to inline it into a view. So it doesn't look like compile is "wrong" here.

**Comment by zou3519** [comment]
I think what's going on here is that Dynamo decided to graph break because the autograd.Function doesn't have a backward and we are tracing into the forward

**Comment by aorenste** [comment]
Classified as a **framework-internal design bug** — the issue describes an architectural problem in PyTorch (dispatcher, codegen, ABI, etc.); any runnable snippet is illustrative of a symptom rather than a reproduction. Auto-triage is skipping this because "no longer reproduces" wouldn't be a valid close signal for an issue like this — the symptom path may have been worked around without fixing the root cause.

Classifier notes: Architectural question about how Dynamo/AOTAutograd handles custom autograd.Function: user observes that inside torch._functorch._aot_autograd.collect_metadata_analysis the grad_fn is None and output_type becomes alias_of_input instead of custom_function_view. zou3519 explained Dynamo graph-breaks because the autograd.Function has no backward and traces into the forward; the snippet is illustrative of the framework-internal behavior, not an error reproduction.

_(Analysis done by Claude.)_
